[Bug]: Cannot build media driver with GCC 16.1.0

### Which component impacted?

Build

### Is it regression? Good in old configuration?

No, this issue exist a long time

### What happened?

While building intel media driver with GCC 16.1.0, the compiler detects the bug:
```
[ 24%] Building CXX object media_driver/CMakeFiles/iHD_drv_video_CODEC.dir/agnostic/common/codec/hal/codechal_encode_hevc_base.cpp.o
/q/media_driver/agnostic/gen8/hw/mhw_state_heap_g8.c: In member function 'virtual MOS_STATUS MHW_STATE_HEAP_INTERFACE_G8_X::AddSamplerStateData(uint32_t, MemoryBlock*, PMHW_SAMPLER_STATE_PARAM)':
/q/media_driver/agnostic/gen8/hw/mhw_state_heap_g8.c:834:38: error: storing the address of local variable 'indirectState' in '*pParam._MHW_SAMPLER_STATE_PARAM::<anonymous>._MHW_SAMPLER_STATE_PARAM::<unnamed union>::Unorm._MHW_SAMPLER_STATE_UNORM_PARAM::pIndirectState' [-Werror=dangling-pointer=]
  834 |         pParam->Unorm.pIndirectState = &indirectState;
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
/q/media_driver/agnostic/gen8/hw/mhw_state_heap_g8.c:832:57: note: 'indirectState' declared here
  832 |         mhw_state_heap_g8_X::SAMPLER_INDIRECT_STATE_CMD indirectState;
      |                                                         ^~~~~~~~~~~~~
/q/media_driver/agnostic/gen8/hw/mhw_state_heap_g8.c:822:33: note: 'pParam' declared here
  822 |     PMHW_SAMPLER_STATE_PARAM    pParam)
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
cc1plus: all warnings being treated as errors
gmake[2]: *** [media_driver/CMakeFiles/iHD_drv_video_COMMON.dir/build.make:1143: media_driver/CMakeFiles/iHD_drv_video_COMMON.dir/agnostic/gen8/hw/mhw_state_heap_g8.c.o] Error 1
```

The issue is presented in latest master (d3c30f6dd), releases 24.4.4, 26.2.1, 26.1.2

### What's the usage scenario when you are seeing the problem?

Others

### What impacted?

Using latest release of GCC: GCC 16.1.0 from official gcc docker image

### Debug Information

1. GCC 16.1.0 docker container default (installed via apt)
2-6. Not actual

### Do you want to contribute a patch to fix the issue?

No.

P.S. Looks like it's orthogonal for issue #1981 - in my case I cannot finalize the driver compiling at all

--- a/media_driver/agnostic/Xe_M/Xe_XPM/hw/mhw_state_heap_xe_xpm.c
+++ b/media_driver/agnostic/Xe_M/Xe_XPM/hw/mhw_state_heap_xe_xpm.c
@@ -178,7 +178,18 @@ MOS_STATUS MHW_STATE_HEAP_INTERFACE_XE_XPM::AddSamplerStateData(
         mhw_state_heap_xe_xpm::SAMPLER_STATE_CMD          unormSampler;
         mhw_state_heap_xe_xpm::SAMPLER_INDIRECT_STATE_CMD indirectState;
 
+        // indirectState is scratch space consumed entirely within this block:
+        // SetSamplerState writes to it via pParam->Unorm.pIndirectState and the
+        // result is copied out below. The pointer never outlives indirectState,
+        // so suppress GCC's false-positive dangling-pointer warning (GCC 16+).
+#if defined(__GNUC__) && !defined(__clang__) && (__GNUC__ >= 12)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdangling-pointer"
+#endif
         pParam->Unorm.pIndirectState = &indirectState;
+#if defined(__GNUC__) && !defined(__clang__) && (__GNUC__ >= 12)
+#pragma GCC diagnostic pop
+#endif
 
         MHW_MI_CHK_STATUS(SetSamplerState(&unormSampler, pParam));
 

--- a/media_driver/agnostic/gen11/hw/mhw_state_heap_g11.c
+++ b/media_driver/agnostic/gen11/hw/mhw_state_heap_g11.c
@@ -894,7 +894,18 @@ MOS_STATUS MHW_STATE_HEAP_INTERFACE_G11_X::AddSamplerStateData(
         mhw_state_heap_g11_X::SAMPLER_STATE_CMD          unormSampler;
         mhw_state_heap_g11_X::SAMPLER_INDIRECT_STATE_CMD indirectState;
 
+        // indirectState is scratch space consumed entirely within this block:
+        // SetSamplerState writes to it via pParam->Unorm.pIndirectState and the
+        // result is copied out below. The pointer never outlives indirectState,
+        // so suppress GCC's false-positive dangling-pointer warning (GCC 16+).
+#if defined(__GNUC__) && !defined(__clang__) && (__GNUC__ >= 12)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdangling-pointer"
+#endif
         pParam->Unorm.pIndirectState = &indirectState;
+#if defined(__GNUC__) && !defined(__clang__) && (__GNUC__ >= 12)
+#pragma GCC diagnostic pop
+#endif
 
         MHW_MI_CHK_STATUS(SetSamplerState(&unormSampler, pParam));
 

--- a/media_driver/agnostic/gen12/hw/mhw_state_heap_g12.c
+++ b/media_driver/agnostic/gen12/hw/mhw_state_heap_g12.c
@@ -777,7 +777,18 @@ MOS_STATUS MHW_STATE_HEAP_INTERFACE_G12_X::AddSamplerStateData(
         mhw_state_heap_g12_X::SAMPLER_STATE_CMD          unormSampler;
         mhw_state_heap_g12_X::SAMPLER_INDIRECT_STATE_CMD indirectState;
 
+        // indirectState is scratch space consumed entirely within this block:
+        // SetSamplerState writes to it via pParam->Unorm.pIndirectState and the
+        // result is copied out below. The pointer never outlives indirectState,
+        // so suppress GCC's false-positive dangling-pointer warning (GCC 16+).
+#if defined(__GNUC__) && !defined(__clang__) && (__GNUC__ >= 12)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdangling-pointer"
+#endif
         pParam->Unorm.pIndirectState = &indirectState;
+#if defined(__GNUC__) && !defined(__clang__) && (__GNUC__ >= 12)
+#pragma GCC diagnostic pop
+#endif
 
         MHW_MI_CHK_STATUS(SetSamplerState(&unormSampler, pParam));
 

--- a/media_driver/agnostic/gen8/hw/mhw_state_heap_g8.c
+++ b/media_driver/agnostic/gen8/hw/mhw_state_heap_g8.c
@@ -831,7 +831,18 @@ MOS_STATUS MHW_STATE_HEAP_INTERFACE_G8_X::AddSamplerStateData(
         mhw_state_heap_g8_X::SAMPLER_STATE_CMD          unormSampler;
         mhw_state_heap_g8_X::SAMPLER_INDIRECT_STATE_CMD indirectState;
 
+        // indirectState is scratch space consumed entirely within this block:
+        // SetSamplerState writes to it via pParam->Unorm.pIndirectState and the
+        // result is copied out below. The pointer never outlives indirectState,
+        // so suppress GCC's false-positive dangling-pointer warning (GCC 16+).
+#if defined(__GNUC__) && !defined(__clang__) && (__GNUC__ >= 12)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdangling-pointer"
+#endif
         pParam->Unorm.pIndirectState = &indirectState;
+#if defined(__GNUC__) && !defined(__clang__) && (__GNUC__ >= 12)
+#pragma GCC diagnostic pop
+#endif
 
         MHW_MI_CHK_STATUS(SetSamplerState(&unormSampler, pParam));
 

--- a/media_driver/agnostic/gen9/hw/mhw_state_heap_g9.c
+++ b/media_driver/agnostic/gen9/hw/mhw_state_heap_g9.c
@@ -1035,7 +1035,18 @@ MOS_STATUS MHW_STATE_HEAP_INTERFACE_G9_X::AddSamplerStateData(
         mhw_state_heap_g9_X::SAMPLER_STATE_CMD          unormSampler;
         mhw_state_heap_g9_X::SAMPLER_INDIRECT_STATE_CMD indirectState;
 
+        // indirectState is scratch space consumed entirely within this block:
+        // SetSamplerState writes to it via pParam->Unorm.pIndirectState and the
+        // result is copied out below. The pointer never outlives indirectState,
+        // so suppress GCC's false-positive dangling-pointer warning (GCC 16+).
+#if defined(__GNUC__) && !defined(__clang__) && (__GNUC__ >= 12)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdangling-pointer"
+#endif
         pParam->Unorm.pIndirectState = &indirectState;
+#if defined(__GNUC__) && !defined(__clang__) && (__GNUC__ >= 12)
+#pragma GCC diagnostic pop
+#endif
 
         MHW_MI_CHK_STATUS(SetSamplerState(&unormSampler, pParam));
 

